distributed/core.py

```diff
@@ -1564,43 +1564,75 @@ async def connect(self, addr: str, timeout: float | None = None) -> Comm:
         if self.semaphore.locked():
             self.collect()
 
-        # This construction is there to ensure that cancellation requests from
-        # the outside can be distinguished from cancellations of our own.
-        # Once the CommPool closes, we'll cancel the connect_attempt which will
-        # raise an OSError
-        # If the ``connect`` is cancelled from the outside, the Event.wait will
-        # be cancelled instead which we'll reraise as a CancelledError and allow
-        # it to propagate
-        connect_attempt = asyncio.create_task(self._connect(addr, timeout))
-        done = asyncio.Event()
-        connecting = self._connecting[addr]
-        connecting.add(connect_attempt)
-
-        def callback(task: asyncio.Task[Comm]) -> None:
-            done.set()
+        if sys.version_info >= (3, 11):
+            reason: str | None = None
+            try:
+                async with asyncio.timeout(math.inf) as scope:
```

Comment:
All of this feels quite magical, could you add a comment on what/why we're doing it?